### PR TITLE
[QMS-158] Change Routino Profiles search for [prefix-]profiles.xml

### PR DESCRIPTION
**Pull request by jmfloyd**
**What is the linked issue for this pull request (start with a `#`):** QMS-#158

**Describe roughly what you have done:**

I have added extra search paths for databasepath/[<prefix>-]profiles.xml and default_ routino_data_path/profiles.xml. Similar to what routino cammandline does.  This is done in the buildDatabaseList method.  This has required moving the parseProfilesXML call into a new method loadProfiles() which is called from the calcRoute methods and expanding the userData argument to a map variable containing the database pointer and the path to an existing profiles.xml.

**What steps have to be done to perform a simple smoke test:**

1. Run routing without any profiles.xml in the database folder. Program uses profiles.xml from the default routino path.
2. Create a prefix-profiles.xml in the database folder with slightly different parameters, and it routes with those characterisitcs.
3. Change the routing database in the UI and routing with a new profiles works.
4. Tested a profiles.xml file with inserted bad data. This created an error message and abort as expected with  from the profiles parse call. 
5. Have been running this on a project where I setup a walking/foot only database and profiles, along with a motorcar only database and profiles. Databases and profiles use unique prefixes.  Can switch seamlessly between databases in one session.

**Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):**

- [x] yes

**Is every user facing string in a tr() macro?**

- [x] yes

**Is the change user facing?**

- [x] yes,
- [ ] no

**If there are user facing changes did you add the ticket number and title into the changelog?**

- [x] yes, I didn't forget to change changelog.txt

### CODE REVIEW

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 1]
```diff
@@ -248,19 +245,49 @@ void CRouterRoutino::buildDatabaseList()
...
 #else
             Routino_Database * data = Routino_LoadDatabase(dir.absolutePath().toUtf8(), prefix.toUtf8());
 #endif
-            if(data)
+            qDebug() << "Loaded Routino DB" << dir.absolutePath().toUtf8().data() << "  " << prefix.toUtf8().data();
+
+            if(data == nullptr)
             {
-                comboDatabase->addItem(prefix.replace("_", " "), quint64(data));
+                QMessageBox::critical(this, "Routino Load...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterRoutino.cpp:259]: Missing tr macro `Routino Load ...`. This has to be translated.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 2]
```diff
@@ -269,12 +296,29 @@ void CRouterRoutino::freeDatabaseList()
...
+    // load xml profiles if it is new before setting the profile!
+    if( currentProfilesPath != profilesPath)
+    {
+        currentProfilesPath = profilesPath;
+        int res = 0;
+        res = Routino_ParseXMLProfiles(profilesPath.toStdString().data());
+        if(res)
+        {
+            QMessageBox::critical(this, "Routino...", xlateRoutinoError(Routino_errno)+"\n '"+profilesPath+"'", QMessageBox::Abort);
+            abort();
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterRoutino.cpp:317]: Abort execution and create a coredump? That's a quite drastic way to handle that kind of error.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 3]
```diff
@@ -269,12 +296,29 @@ void CRouterRoutino::freeDatabaseList()
...
 }
 
+void CRouterRoutino::loadProfiles(QString profilesPath)
+{
+    // load xml profiles if it is new before setting the profile!
+    if( currentProfilesPath != profilesPath)
+    {
+        currentProfilesPath = profilesPath;
+        int res = 0;
+        res = Routino_ParseXMLProfiles(profilesPath.toStdString().data());
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterRoutino.cpp:313]: please use `toUtf8()` as in the original call.

File: changelog.txt [hunk 4]
```diff
@@ -14,6 +14,7 @@ V1.XX.X
 [QMS-173] Misleading help text for DEM range slider
 [QMS-178] Automate list of code contributors in about dialog
 [QMS-183] Backward range selection causing small issues
+[QMS-158] Change Routino Profiles search for [prefix-]profiles.xml
```
Comment (kiozen) [inline changelog.txt:17]: Please keep the number in order.

**Comment by kiozen** [comment]
any progress here?

**Comment by jmfloyd** [comment]
Thanks.  Will be getting back to this very soon.  Been very busy lately.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 5]
```diff
@@ -248,19 +245,49 @@ void CRouterRoutino::buildDatabaseList()
...
+            dmap["profilesPath"] = "";
+            QStringList profilesPaths = {
+                dir.filePath(prefix+"-profiles.xml"),
+                dir.filePath("profiles.xml"),
+                setup->routinoPath("profiles.xml").toUtf8()
+            };
+            for(const QString& profilePath : profilesPaths)
             {
-                QMessageBox::critical(this, "Routino...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
+                QFileInfo pinfo = QFileInfo(profilePath);
```
[the remark this answers is not in the document]
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:275]: Thanks - I forget that the class has an implicit new() method. Done

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 6]
```diff
@@ -248,19 +245,49 @@ void CRouterRoutino::buildDatabaseList()
...
+                if( pinfo.isReadable())
+                {
+                    dmap["profilesPath"] = pinfo.filePath();
+                    break;
+                }
+            }
+            if( dmap["profilesPath"].toString().length() == 0)
+            {
+                QMessageBox::critical(this, "Routino...",tr("Could not find a profiles XML file in expected folders. Routing will not function"), QMessageBox::Ok);
```
[the remark this answers is not in the document]
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:284]: Why I left it like this is that it reports the error but is does not cause any segmentation errors further on - just no action.  I thought that throwing you out of the program is a bit severe.  Other routing options are available without aborting.  Suggestions welcomed.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 7]
```diff
@@ -269,12 +296,29 @@ void CRouterRoutino::freeDatabaseList()
...
     {
-        Routino_Database * data = (Routino_Database*)comboDatabase->itemData(i, Qt::UserRole).toULongLong();
+        QVariantMap map = comboDatabase->itemData(i,Qt::UserRole).toMap();
+        Routino_Database * data = (Routino_Database*)(map["db"].toULongLong());
         Routino_UnloadDatabase(data);
     }
     comboDatabase->clear();
 }
 
+void CRouterRoutino::loadProfiles(QString profilesPath)
```
[the remark this answers is not in the document]
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:306]: Thanks true and done

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 1 above)
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:259]: Thanks missed that.  Changed it back to just "Routino ..."  which is used is other places.  So no tr() required

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 2 above)
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:317]: This was done because it leads to seg errors in the routing calculations.  These are not wanted.  The only way I can see to get around this is to test load the profiles  in the buildDatabaseList() call.  This is being implemented.  In the mean time leave this abort here but should never be called if the errors are caught earlier.  Suggestions welcomed.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 8]
```diff
@@ -248,19 +245,49 @@ void CRouterRoutino::buildDatabaseList()
...
             {
-                QMessageBox::critical(this, "Routino...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
+                QFileInfo pinfo = QFileInfo(profilePath);
+                if( pinfo.isReadable())
+                {
+                    dmap["profilesPath"] = pinfo.filePath();
+                    break;
+                }
+            }
+            if( dmap["profilesPath"].toString().length() == 0)
```
[the remark this answers is not in the document]
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:282]: Done

**Comment by jmfloyd** [comment]
Moved into sequential order ..

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 2 above)
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:317]: OK I now have loadProfiles called during the buildDatabaseList which catches any profile errors.  But what is the recommended way to smoothly exit the program without coredumps etc - use throw QString() also does a coredump??  What do you recommend for the button to close the window, I would have thought OK if I add extra text to indicate the  file needs fixing?  Please your suggestion.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kiozen) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp]: Using QMessageBox::Abort instead of QMessageBox::Ok will not terminate anything. It's simply about naming the button. And aborting to load the particular profile for that particular database is what you are doing. It's also good pracatice to mention the database you are aborting to load and configure.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 2 above)
Comment (kiozen) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp]: Never terminate the program the hard way. You will destroy the users unsaved data. No user will appreciate that. This is not a command line application you can abort on an error to fall back to command line. This is a very complex GUI application that should not be closed by no one else than the user by intention. Everything else is considered a bug. 

That in mind: Inform the user that something is wrong. Give as much information to help the user to fix it. After all we are in that condition because the user did something peculiar. It's also ok to add a QLabel to the GUI that tells in red fat letters that routing is not available due to errors. BRouter has such a label. Additionally you have to block test in the API of CRouterRoutino if routing is possible. If not simply return with no result. 

In other words: In the error case every thing should work as expected. Just the faulty function is not working.Ideally there is some feedback to inform the user.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 3 above)
Comment (kiozen) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp]: Ok, but remove the commented line of code.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp [hunk 9]
```diff
@@ -248,19 +245,54 @@ void CRouterRoutino::buildDatabaseList()
...
             {
-                comboDatabase->addItem(prefix.replace("_", " "), quint64(data));
+                QMessageBox::critical(this, "Routino...",tr("Could not find a profiles XML file in expected folders. Routino Routing will not function"), QMessageBox::Ok);
+                return;
             }
             else
             {
-                QMessageBox::critical(this, "Routino...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
+                /* check if profile will load - will abort if not good */
+                loadProfiles(dmap["profilesPath"].toString());
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterRoutino.cpp:290]: `loadProfiles()` might throw an exception that is not caught -> app crash -> user looses data.

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 3 above)
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp]: Done

File: changelog.txt (hunk 4 above)
Comment (jmfloyd) [reply changelog.txt]: Done

File: src/qmapshack/gis/rte/router/CRouterRoutino.cpp (hunk 9 above)
Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:290]: Understand that.  Will investigate a way to avoid an abort.

Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:290]: I propose to check the profiles file during builddatabaselist and if an error occurs, do not add  that database to the list (so it is not accessable) with a warning to fix the xml file.  I have tried this  and if there are multiple databases the good ones are still available but the faulty one is not.  This avoids any segmentation due to that file.

Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp:290]: [Routino _002](https://user-images.githubusercontent.com/1490471/84631344-8f8e9a80-af30-11ea-8385-05de742998ba.png)

Here is my error message for a corrupted profiles.xml.  It does not abort, but the database is put not in the combo list. So it cant be used.

Comment (jmfloyd) [reply src/qmapshack/gis/rte/router/CRouterRoutino.cpp]: I have now shifted things around.  loadProfiles is called in buildDatabaseList.  If an error occurs, then that database is not included in the databases able to be used.  A warning messagebox is emitted, indicating that the profiles file needs fixing and that the database is not available to use.  No aborts or throws are generated. loadProfiles is still called as required in the routing routines to use the correct profile parameters.

**Comment by kiozen** [comment]
You have closed the PR?

**Comment by jmfloyd** [comment]
> You have closed the PR? 
No - but due to some lingering git problems caused by me, I recreated an new empty QMS-158 branch in the forked repository.  It appeared just minutes after the original was deleted.

**Comment by jmfloyd** [comment]
OK does this mean I need to post another PR?  I had thought that was always the case whenever I had a new updated version available.

**Comment by kiozen** [comment]
Now you have to because the branch has been removed.

But for the future: You one PR is enough :wink: You commit your changes either by appending the las commit and push it (use option `--force-with-lease`). Or you do another commit if that is appropriate. `appropriate` means: The changes of the commit are kind of unrelated to the previous commits, e.g. they affect another group of files or introduce another functionality on top of other commits.

And you have rebase your branch if guthub tells you that it can't be merged. This is done by:
```
git checkout dev
git fetch upstream 
git reset upstream/dev
git checkout <your branch>
git rebase dev
```
The rebase might be stopped due to conflicts. You have to resolve the conflicts by merging or removing the marked code. I prefer to use Visual Code for that task.

The rebase command is also a very good tool to clean up your commit history like merge commits together or rephrasing commit messages. With `git log` you can see your commit history. Let's asume you have 3 commits and want to edit them. You do:

```
git rebase -i HEAD~3
```

Now you see a list with those three commits. Read the documentation provided by git to get a clue what to do. It's a very powerful and flexible command once you get used to it.

When done and everything is fine you do

```
git push --force-with-lease
```

and the PR will be updated.
